rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/History.scala

```diff
@@ -0,0 +1,375 @@
+                pointerBlock
+                  .updated((toInt(keyLast), EmptyTrie) :: Nil)
+                  .toVector
+                  .zipWithIndex
+                  .filter(v => v._1 != EmptyTrie)
+                  .head
+
+              val (other, idx) = getOther
+              Applicative[F].pure((keyInit, init, Some(pointTo(toByte(idx), other)))).map(_.asLeft)
+            case Some(pb) => // pb contains 3+ elements, drop one, rehash
+              val updated = pb.updated((toInt(keyLast), EmptyTrie) :: Nil)
+              storePointerBlock(updated)
+                .flatMap(node => rehash(keyInit, init, node, Nil).map(_.asRight))
+            case None => Sync[F].raiseError(new RuntimeException("malformed trie"))
+          }
+
+        case (key, rest @ _ :+ (_: Node), Some(s: Skip)) =>
+          rehash(key, rest, s, Nil).map(_.asRight)
+
+        case (Nil, Vector(), Some(t)) => Applicative[F].pure(t :: Nil).map(_.asRight)
+
+        case (Nil, Vector(), None) => Applicative[F].pure(EmptyTrie :: Nil).map(_.asRight)
+      }
+    FlatMap[F].tailRecM((p1, r, a))(go)
+  }
+
+  def process(actions: List[HistoryAction[K]]): F[History[F, K]] =
+    // TODO this is an intermediate step to reproduce all the trie behavior
+    // will evolve to a fold based implementation with partial tries
+    actions
+      .foldLeftM(this.root) {
+        case (currentRoot, InsertAction(k, value)) =>
+          val currentPath = asBytes(k)
+          for {
+            traverseResult <- findPath(currentPath, currentRoot)
+            (_, triePath)  = traverseResult
+            elements       <- rebalanceInsert(Leaf(value), currentPath, triePath)
+            _              <- historyStore.put(elements)
+          } yield Trie.hash(elements.last)
+
+        case (currentRoot, DeleteAction(k)) =>
+          val currentPath = asBytes(k)
+          for {
+            traverseResult <- findPath(currentPath, currentRoot)
+            (_, triePath)  = traverseResult
+            elements <- triePath match {
+                         case TriePath(elems :+ (_: Leaf), None, path) if currentPath == path =>
+                           rebalanceDelete(currentPath, elems, None)
+                         case _ =>
+                           Applicative[F].pure[List[Trie]](Nil)
+                       }
+            _ <- historyStore.put(elements)
+          } yield elements.lastOption.map(Trie.hash).getOrElse(currentRoot)
+      }
+      .map(newRoot => this.copy(root = newRoot))
+
+  private[history] def findPath(key: KeyPath, start: Blake2b256Hash): F[(Trie, TriePath)] = {
+    type Params = (Trie, List[Byte], TriePath)
+    //t: Trie, currentPath: List[Byte], path: TriePath
+    def traverse(params: Params): F[Either[Params, (Trie, TriePath)]] =
+      params match {
+        case (EmptyTrie, _, path) => Applicative[F].pure((EmptyTrie, path)).map(_.asRight)
+        case (l: Leaf, _, path)   => Applicative[F].pure((l, path append (Nil, l))).map(_.asRight)
```

Comment:
I think we should be requiring an empty `key` / `remainingPath` (middle tuple element) in the first two cases. Otherwise we've lost track during navigation and it's a malformed tree / bug

Reply:
because an empty trie is EmptyTrie. it's an edge case.

Reply:
https://rchain.atlassian.net/browse/RCHAIN-3228

Reply:
My point was I don't see a difference in terms of path in traversing to a Leaf vs traversing to an Empty. I'd expect both cases to enlarge the path.